### PR TITLE
Changes to support TP-4068  and TP-4084

### PR DESCRIPTION
**Pull request by jeffreyalien**
TP-4068 Contains changes to support Enhanced Protection Information as well as support for extended Reference Tags
TP-4084 Contains changes to support Time to Ready enhancements.

### CODE REVIEW

**Review by bjpaupor — APPROVED** [review]

File: nvme-print.c [hunk 1]
```diff
@@ -2925,22 +2941,23 @@ static void json_ctrl_registers(void *bar)
 	pmrmscu = mmio_read32(bar + NVME_REG_PMRMSCU);
 
 	root = json_create_object();
-	json_object_add_value_uint(root, "cap", cap);
+	json_object_add_value_uint64(root, "cap", cap);
```
Comment (igaw) [inline nvme-print.c:2944]: Good catch, but these fixes should be in a separate patch.

File: nvme-print.c [hunk 2]
```diff
@@ -2608,6 +2612,9 @@ static void nvme_show_registers_cc_shn (__u8 shn)
 
 static void nvme_show_registers_cc(__u32 cc)
 {
+	printf("\tController Ready Independent of Media Enable (CRIME): %s\n",
+		(cc & 0x01000000) ? "Enabled":"Disabled");
```
Comment (igaw) [inline nvme-print.c:2616]: Why not using the newly introduced NVME_CC_CRIME() macro from libnvme?

File: nvme-print.c [hunk 3]
```diff
@@ -2656,6 +2663,14 @@ static void nvme_show_registers_csts(__u32 csts)
 
 }
 
+static void nvme_show_registers_crto(__u32 crto)
+{
+	printf("\tCRIMT                               : %d secs\n",
+		((crto & NVME_CRTO_CRIMT_MASK) >> NVME_CRTO_CRIMT_SHIFT)/2 );
+	printf("\tCRWMT                               : %d secs\n",
+		((crto & NVME_CRTO_CRWMT_MASK) >> NVME_CRTO_CRWMT_SHIFT)/2 );
```
Comment (igaw) [inline nvme-print.c:2671]: and here  `NVME_CRTO_CRWMT(c)/2`?

File: nvme.c [hunk 4]
```diff
@@ -5398,10 +5399,47 @@ static int write_uncor(int argc, char **argv, struct command *cmd, struct plugin
...
+static int invalid_tags(__u64 storage_tag, __u64 ref_tag, __u8 sts, __u8 pif)
+{
+	int result = 0;
+	if (sts < 64 && storage_tag >= (1LL << sts)) {
+		fprintf(stderr, "Storage tag larger than storage tag size\n");
+		return 1;
+	}
+
+	switch (pif)
+	{
```
Comment (igaw) [inline nvme.c:5411]: the `{` goes to the line above `switch (pif) {`

File: subprojects/libnvme.wrap [hunk 5]
```diff
@@ -1,6 +1,6 @@
 [wrap-git]
 url = https://github.com/linux-nvme/libnvme.git
-revision = 5ba1e594141d6f3fbbb1efac97c4865fa67571d3
+revision = 91f66d582c1f107971042afa57614f08f0fc6d76
```
Comment (igaw) [inline subprojects/libnvme.wrap:3]: please move this into a separate commit, so we do mix build changes with code changes.

File: nvme.c [hunk 6]
```diff
@@ -5398,10 +5399,47 @@ static int write_uncor(int argc, char **argv, struct command *cmd, struct plugin
 	return err;
 }
 
+static int invalid_tags(__u64 storage_tag, __u64 ref_tag, __u8 sts, __u8 pif)
+{
+	int result = 0;
+	if (sts < 64 && storage_tag >= (1LL << sts)) {
```
Comment (igaw) [inline nvme.c:5405]: new line after the declaration.

**Review by igaw — CHANGES_REQUESTED** [review]
A bunch of nitpicks.

File: nvme-print.c [hunk 7]
```diff
@@ -2608,6 +2612,9 @@ static void nvme_show_registers_cc_shn (__u8 shn)
 
 static void nvme_show_registers_cc(__u32 cc)
 {
+	printf("\tController Ready Independent of Media Enable (CRIME): %s\n",
+		(cc & 0x01000000) ? "Enabled":"Disabled");
+
 	printf("\tI/O Completion Queue Entry Size (IOCQES): %u bytes\n",
 		1 << ((cc & 0x00f00000) >> NVME_CC_IOCQES_SHIFT));
```
[the remark this answers is not in the document]
Comment (jeffreyalien) [reply nvme-print.c:2619]: Since there's a bunch of these types of changes, I'll do them in a separate PR.

File: nvme-print.c (hunk 1 above)
Comment (jeffreyalien) [reply nvme-print.c:2944]: Moved to separate commit - dea1ddb726dc06fdefd12d57e241e8a17c85caf4

File: nvme-print.c [hunk 8]
```diff
@@ -2656,6 +2663,14 @@ static void nvme_show_registers_csts(__u32 csts)
 
 }
 
+static void nvme_show_registers_crto(__u32 crto)
+{
+	printf("\tCRIMT                               : %d secs\n",
+		((crto & NVME_CRTO_CRIMT_MASK) >> NVME_CRTO_CRIMT_SHIFT)/2 );
```
[the remark this answers is not in the document]
Comment (jeffreyalien) [reply nvme-print.c:2669]: Fixed in commit 54569c34737fe22b52ed91da5305d0222ded9eb1

File: subprojects/libnvme.wrap (hunk 5 above)
Comment (jeffreyalien) [reply subprojects/libnvme.wrap:3]: fixed in commit 455f82e12e4ccfc5e4fe18e5ba2b26d700f69d75

**Comment by jeffreyalien** [comment]
@igaw  I fixed all your comments.  Please review and let me know if you find anything else.

**Comment by igaw** [comment]
FTR, I've changed following things
- reordered the patches: libnvme update, bug fixes, new features. 
- small changes in the commit messages

**Review by igaw — APPROVED** [review]
